bal/bal.c

```diff
@@ -76,14 +88,48 @@ baldev_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
 	if (count > BAL_MAX_BUF_SIZE)
 		return -EMSGSIZE;
 
-	status = wait_for_busy_idle();
-	if (0 == status) {
-		status = spi_read(bal.spi, bal.buffer, count);
-	}
+	if(bal.HalType == BAL_HAL_HW_PN5180)
+	{
+		status = wait_for_busy_idle();
 
-	if (copy_to_user(buf, bal.buffer, count)) {
-		return -EFAULT;
+		if (0 == status)
+			status = spi_read(bal.spi, bal.buffer, count);
+	}
+	else
+	{
+
+		status = copy_from_user(bal.buffer, buf, count);
+
+		bal.xfers.tx_buf = bal.buffer;
+		bal.xfers.rx_buf = bal.buffer;
+		bal.xfers.len = count;
+
+		//"Fast" bundled exchange for EMVCo compatibility -- only necessary for PN512/RC663 derivatives
+		if( count < 1 )
+			return count;
+
+	    //This kenel module does not need to check read/write flag bit. It is upon caller do decide whether read or write.
+		if(bal.MultiRegRW == 1)
+	    {
+	    	//Perform a "MultiRegRead" exchange
```

Comment:
There is no difference. I just wanted to keep formal spliting like in LPC project in LPC software package.